### PR TITLE
Spectrum info rollout 13 to 20

### LINKED ISSUE #18233: SpectrumInfo rollout: algorithms 13 to 20
This issue is part of the SpectrumInfo rollout, see https://github.com/mantidproject/mantid/issues/17743.
Algorithms covered in this issue are:
* Framework/Algorithms/src/CalculateDIFC.cpp
* Framework/Algorithms/src/ExtractMaskToTable.cpp
* Framework/Algorithms/src/He3TubeEfficiency.cpp
* Framework/Algorithms/src/SmoothNeighbours.cpp
* Framework/Algorithms/src/SofQWNormalisedPolygon.cpp
* Framework/Algorithms/src/SampleCorrections/MayersSampleCorrection.cpp
* Framework/API/src/MatrixWorkspaceMDIterator.cpp
* Framework/CurveFitting/src/SeqDomainSpectrumCreator.cpp

### PR DESCRIPTION
**Pull request by ianbush**
## Description of work.
This PR is part of refactoring effort to use the new interface of the SpectrumInfo module.
See https://github.com/mantidproject/mantid/issues/17743 for details.

The following algorithms were refactored to use SpectrumInfo:
* Framework/Algorithms/src/CalculateDIFC.cpp
* Framework/Algorithms/src/ExtractMaskToTable.cpp
* Framework/Algorithms/src/He3TubeEfficiency.cpp
* Framework/Algorithms/src/SmoothNeighbours.cpp
* Framework/Algorithms/src/SofQWNormalisedPolygon.cpp
* Framework/Algorithms/src/SampleCorrections/MayersSampleCorrection.cpp
* Framework/API/src/MatrixWorkspaceMDIterator.cpp
* Framework/CurveFitting/src/SeqDomainSpectrumCreator.cpp

Performance tests were added for: N/A

The following algorithms have improved performance: N/A

## To test.

Code review.

Fixes #18233.
Internal change, no release notes (potential performance improvements will be added to release notes as part of the umbrella issue).

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [ ] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?

- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/API/inc/MantidAPI/MatrixWorkspaceMDIterator.h [hunk 1]
```diff
@@ -138,6 +139,9 @@ class DLLExport MatrixWorkspaceMDIterator : public IMDIterator {
   /// For numeric axes, this is the size of the bin in the vertical direction.
   /// It is 1.0 for spectrum axes
   double m_verticalBinSize;
+
+  /// SpectrumInfo object, used for masking information
+  SpectrumInfo m_spectrumInfo;
```
Comment (SimonHeybrock) [inline Framework/API/inc/MantidAPI/MatrixWorkspaceMDIterator.h]: `const &`?

File: Framework/Algorithms/inc/MantidAlgorithms/He3TubeEfficiency.h [hunk 2]
```diff
@@ -84,11 +85,11 @@ class DLLExport He3TubeEfficiency : public API::Algorithm {
...
 
   /// Correct the given spectra index for efficiency
-  void correctForEfficiency(std::size_t spectraIndex);
+  void correctForEfficiency(std::size_t spectraIndex,
+                            const API::SpectrumInfo &spectrumInfo);
   /// Sets the detector geometry cache if necessary
-  void
-  getDetectorGeometry(const boost::shared_ptr<const Geometry::IDetector> &det,
-                      double &detRadius, Kernel::V3D &detAxis);
+  void getDetectorGeometry(const Geometry::IDetector &det, double &detRadius,
```
Comment (SimonHeybrock) [inline Framework/Algorithms/inc/MantidAlgorithms/He3TubeEfficiency.h]: 👍

File: Framework/Algorithms/src/CalculateDIFC.cpp [hunk 3]
```diff
@@ -94,24 +91,24 @@ void CalculateDIFC::calculate(API::Progress &progress,
                               DataObjects::OffsetsWorkspace_sptr &offsetsWS,
                               double l1, double beamlineNorm,
                               Kernel::V3D &beamline, Kernel::V3D &samplePos,
-                              detid2det_map &allDetectors) {
+                              const API::DetectorInfo &detectorInfo) {
   SpecialWorkspace2D_sptr localWS =
       boost::dynamic_pointer_cast<SpecialWorkspace2D>(outputWs);
 
+  const auto detectorIDs = detectorInfo.detectorIDs();
```
Comment (SimonHeybrock) [inline Framework/Algorithms/src/CalculateDIFC.cpp]: `&`

File: Framework/Algorithms/src/ExtractMaskToTable.cpp [hunk 4]
```diff
@@ -202,22 +203,19 @@ std::vector<detid_t> ExtractMaskToTable::extractMaskFromMatrixWorkspace() {
...
-  Instrument_const_sptr instrument = m_dataWS->getInstrument();
-  if (!instrument)
+  const auto &detectorInfo = m_dataWS->detectorInfo();
+  if (detectorInfo.size() == 0)
     throw runtime_error("There is no instrument in input workspace.");
 
   // Extract
-  size_t numdets = instrument->getNumberDetectors();
-  vector<detid_t> detids = instrument->getDetectorIDs();
+  vector<detid_t> detids = detectorInfo.detectorIDs();
```
Comment (SimonHeybrock) [inline Framework/Algorithms/src/ExtractMaskToTable.cpp]: `&`

File: Framework/Algorithms/src/SampleCorrections/MayersSampleCorrection.cpp [hunk 5]
```diff
@@ -101,25 +100,25 @@ void MayersSampleCorrection::exec() {
...
 
     MayersSampleCorrectionStrategy::Parameters params;
     params.mscat = mscatOn;
-    params.l1 = l1;
-    params.l2 = det->getDistance(*sample);
-    params.twoTheta = det->getTwoTheta(sample->getPos(), beamLine);
-    params.phi = det->getPhi();
+    params.l1 = spectrumInfo.l1();
+    params.l2 = spectrumInfo.l2(i);
+    params.twoTheta = spectrumInfo.twoTheta(i);
```
Comment (SimonHeybrock) [inline Framework/Algorithms/src/SampleCorrections/MayersSampleCorrection.cpp]: Doesn't this do something else?

File: Framework/Algorithms/src/SofQWNormalisedPolygon.cpp [hunk 6]
```diff
@@ -354,9 +363,9 @@ void SofQWNormalisedPolygon::initAngularCachesPSD(
       g_log.debug() << "Neighbor ID: " << spec << '\n';
       if (spec == deltaPlus1 || spec == deltaMinus1 || spec == deltaPlusT ||
           spec == deltaMinusT) {
-        DetConstPtr detector_n = workspace->getDetector(spec - 1);
-        double theta_n = workspace->detectorTwoTheta(*detector_n) * 0.5;
-        double phi_n = detector_n->getPhi();
+        const auto &detector_n = spectrumInfo.detector(spec - 1);
+        double theta_n = workspace->detectorTwoTheta(detector_n) * 0.5;
```
Comment (SimonHeybrock) [inline Framework/Algorithms/src/SofQWNormalisedPolygon.cpp]: `spectrumInfo.twoTheta()`?

**Review by SimonHeybrock — COMMENTED** [review]
Looks good, just minor comments:

File: Framework/Algorithms/src/SampleCorrections/MayersSampleCorrection.cpp (hunk 5 above)
Comment (ianbush) [inline Framework/Algorithms/src/SampleCorrections/MayersSampleCorrection.cpp]: Yes, you are right, this is something slightly different. I have reverted this.

**Comment by SimonHeybrock** [comment]
:shipit:

**Pull request merged**